refactor(core): Move more code into `@n8n/permissions`. Add aditional tests and docs (no-changelog)

netroy opened the pull request:
## Summary

A lot of RBAC code is that should have been abstracted away in `@n8n/permissions` seems to have leaked into other places in the codebase.
This PR is a Part-1 in refactoring this code, and adding tests & jsdocs.

In part-2 we'll delete the `/rest/roles` endpoint, and significantly simplify a lot of RBAC code in the frontend, by directly using the `@n8n/permissions` package, that it already depends on.

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

netroy committed:
refactor(core): Move more code into @n8n/permissions. Add aditional tests (no-changelog)

netroy committed:
Merge remote-tracking branch 'origin/master' into move-more-permissions-code-out-of-cli-part-1

despairblue commented on packages/@n8n/permissions/src/types.ee.ts:
```diff
@@ -1,53 +1,81 @@
 
-export type GlobalRole = 'global:owner' | 'global:admin' | 'global:member';
-export type AssignableRole = Exclude<GlobalRole, 'global:owner'>;
+export type ApiKeyScope = ApiKeyResourceScope<PublicApiKeyResources>;
```
This widens the scope by combining all resources with all operations, as mentioned in the comment in the removed code. E.g. this allows `workflow:changeRole`. 

![image](https://github.com/user-attachments/assets/afc982e1-4904-4ab4-9286-225ec5b78073)

We should bring back the intermediary type.

despairblue replied:
Maybe we should also add a type test that makes sure this is not introduced in the future.

netroy replied:
Thanks for catching this. reverted and added types test.

netroy committed:
revert ApiKeyScope changes and add type-checking tests to prevent this regression in the future.

despairblue committed:
revert Scope changes and add type-checking tests to prevent this regression in the future.

despairblue committed:
Merge remote-tracking branch 'origin/master' into move-more-permissions-code-out-of-cli-part-1

despairblue committed:
fix test not catching regressions

The previous test would have successed with either ScopeOption, `oneOf`
or `allOf`. Thus wasn't sufficient to catch regressions.

despairblue committed:
it's not necessary to test with non-existing scopes when using the member role

thus it's not necessary to use the casting

despairblue committed:
fix test to catch regressions

despairblue committed:
test that the filter actually works

despairblue committed:
test that masking actually works

thanks to mrgeio for catching this

despairblue committed:
add id back to make sure `findOne` is called with the right user later

despairblue commented on packages/cli/src/services/__tests__/credentials-finder.service.test.ts:
```diff
@@ -19,19 +17,10 @@ describe('CredentialsFinderService', () => {
-				hasScope(scope, {
-					global: GLOBAL_MEMBER_SCOPES,
-				}),
+			role: 'global:member',
```
added it back

netroy commented on packages/cli/src/services/role.service.ts:
```diff
@@ -1,159 +1,30 @@
+	getAllRoles() {
+		Object.values(ALL_ROLES).forEach((entries) => {
+			entries.forEach((entry) => {
+				entry.licensed = this.isRoleLicensed(entry.role);
```
this entire code will be removed in part-2, where we also remove the `/roles` endpoint, and use `@n8n/permissions` directly in the UI.

netroy commented on packages/cli/src/controllers/password-reset.controller.ts:
```diff
@@ -62,18 +63,23 @@ export class PasswordResetController {
 
 		// User should just be able to reset password if one is already present
 		const user = await this.userRepository.findNonShellUser(email);
+		if (!user) {
+			this.logger.debug('No user found in the system');
+			return;
+		}
```
did this intentionally to make the rest of the code below less confusing. 
I find negation and optional-chaining in the same expression a bit hard to read.

netroy commented on packages/@n8n/permissions/src/utilities/__tests__/hasScope.test.ts:
I redid the tests to get 100% coverage with just this test, and removed any superfluous tests. 
![image](https://github.com/user-attachments/assets/c35b4d14-c21a-4479-80c9-7c302f22ec22)

netroy committed:
Merge remote-tracking branch 'origin/master' into move-more-permissions-code-out-of-cli-part-1

despairblue committed:
add edge case tests

janober commented:
Got released with [`n8n@1.93.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.93.0)
